Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,589 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Diagnostics.Contracts;
+
+namespace Microsoft.Dafny {
+  public class PreTypeInferenceState {
+    public readonly PreTypeResolver PreTypeResolver;
+    private readonly DafnyOptions options;
+
+    private List<SubtypeConstraint> unnormalizedSubtypeConstraints = new();
+    private List<EqualityConstraint> equalityConstraints = new();
+    private List<Func<bool>> guardedConstraints = new();
+    private readonly List<Advice> defaultAdvice = new();
+    private List<System.Action> confirmations = new();
+
+    public PreTypeInferenceState(PreTypeResolver preTypeResolver) {
+      this.PreTypeResolver = preTypeResolver;
+      this.options = preTypeResolver.resolver.Options;
+    }
+
+    // ---------------------------------------- Pre-type inference state ----------------------------------------
+
+    /// <summary>
+    /// Solves or simplifies as many type constraints as possible.
+    /// If "allowDecisions" is "false", then no decisions, only determined inferences, are made; this mode is
+    /// appropriate for the partial solving that's done before a member lookup.
+    /// Returns "true" if anything changed (that is, if any of the constraints in the type-inference state
+    /// caused a change some pre-type proxy).
+    /// </summary>
+    void PartiallySolveTypeConstraints(string printableContext = null, bool makeDecisions = false) {
+      if (printableContext != null) {
+        PrintTypeInferenceState("(partial) " + printableContext);
+      }
+      bool anythingChanged;
+      do {
+        if (makeDecisions) {
+          if (TryResolveTypeProxiesUsingKnownBounds(true)) {
+            // something changed, so do another round of Apply... calls below
+          } else if (TryResolveTypeProxiesUsingKnownBounds(false)) {
+            // something changed, so do another round of Apply... calls below
+          } else {
+            return;
+          }
+        }
+        anythingChanged = false;
+        anythingChanged |= ApplySubtypeConstraints();
+        anythingChanged |= ApplyEqualityConstraints();
+        anythingChanged |= ApplyGuardedConstraints();
+      } while (anythingChanged);
+    }
+
+    void SolveAllTypeConstraints(string printableContext) {
+      PrintTypeInferenceState(printableContext);
+      PartiallySolveTypeConstraints(null);
+
+      PartiallySolveTypeConstraints(null, true);
+
+      if (TryApplyDefaultAdvice()) {
+        PartiallySolveTypeConstraints(null, true);
+      }
+
+      PrintLegend();
+      ConfirmTypeConstraints();
+      ClearState();
+    }
+
+    void ClearState() {
+      unnormalizedSubtypeConstraints.Clear();
+      equalityConstraints.Clear();
+      guardedConstraints.Clear();
+      defaultAdvice.Clear();
+      confirmations.Clear();
+      PreTypeResolver.allPreTypeProxies.Clear();
+    }
+
+    public void PrintTypeInferenceState(string/*?*/ header = null) {
```

Comment:
I would move all the printing code to the bottom of the file, or otherwise sort the methods topologically.

Reply:
I moved method `DebugPrint` to the end of the file. The others are already sorted topologically, and I think it makes sense to have those methods near where other operations that concern the entire state is placed.